modules/ml/src/main/java/org/apache/ignite/ml/selection/scoring/evaluator/Evaluator.java

```diff
@@ -40,359 +55,707 @@
+                query.iterator().forEachRemaining(kv -> {
+                    LabeledVector vector = preprocessor.apply(kv.getKey(), kv.getValue());
+
+                    for (Class key : aggrs.keySet()) {
+                        MetricStatsAggregator aggr = aggrs.get(key);
+                        aggr.aggregate(mdl, vector);
+                    }
+                });
+            }
 
-        return metricValues;
+            return aggrs;
+        }
+        else {
+            return dataset.compute(data -> {
+                Map<Class, MetricStatsAggregator> aggrs = initAggregators(ctxs, metrics);
+
+                for (int i = 0; i < data.getLabels().length; i++) {
+                    LabeledVector<Double> vector = VectorUtils.of(data.getFeatures()[i]).labeled(data.getLabels()[i]);
+                    for (Class key : aggrs.keySet()) {
+                        MetricStatsAggregator aggr = aggrs.get(key);
+                        aggr.aggregate(mdl, vector);
+                    }
+                }
+
+                return aggrs;
+            }, (left, right) -> {
+                if (left == null && right == null)
+                    return new HashMap<>();
+                if (left == null)
+                    return right;
+                if (right == null)
+                    return left;
+
+                HashMap<Class, MetricStatsAggregator> res = new HashMap<>();
+                for (Class key : left.keySet()) {
+                    MetricStatsAggregator agg1 = left.get(key);
+                    MetricStatsAggregator agg2 = right.get(key);
+                    A.ensure(agg1 != null && agg2 != null, "agg1 != null && agg2 != null");
+                    res.put(key, agg1.mergeWith(agg2));
+                }
+                return res;
+            });
+        }
+    }
+
+    /**
+     * Inits aggregators.
+     *
+     * @param ctxs Evaluation contexts.
+     * @param metrics Metrics.
+     * @return Aggregators map.
+     */
+    private static Map<Class, MetricStatsAggregator> initAggregators(Map<Class, EvaluationContext> ctxs, Metric[] metrics) {
+        Map<Class, MetricStatsAggregator> aggrs = new HashMap<>();
+        for (Metric m : metrics) {
+            MetricStatsAggregator aggregator = m.makeAggregator();
+            EvaluationContext ctx = ctxs.get(aggregator.getClass());
+            A.ensure(ctx != null, "ctx != null");
+            aggregator.initByContext(ctx);
+
+            if (!aggrs.containsKey(aggregator.getClass()))
+                aggrs.put(aggregator.getClass(), aggregator);
+        }
+        return aggrs;
+    }
+
+    /**
+     * Returns true if model should be evaluated only through scan query.
+     *
+     * @param mdl Model to estimation.
+     * @return true if model should be evaluated only through scan query.
+     */
+    private static boolean isOnlyLocalEstimation(IgniteModel<Vector, Double> mdl) {
+        return mdl instanceof KNNModel;
```

Comment:
Does it make sense to make it a property of a model?

Reply:
Maybe, when I'll start work on model meta-information, I'll add this property to model